synapse/config/metrics.py

```diff
@@ -73,46 +73,5 @@ def read_config(self, config: JsonDict, **kwargs: Any) -> None:
     def generate_config_section(
         self, report_stats: Optional[bool] = None, **kwargs: Any
     ) -> str:
-        res = """\
-        ## Metrics ###
-
-        # Enable collection and rendering of performance metrics
-        #
-        #enable_metrics: false
-
-        # Enable sentry integration
-        # NOTE: While attempts are made to ensure that the logs don't contain
-        # any sensitive information, this cannot be guaranteed. By enabling
-        # this option the sentry server may therefore receive sensitive
-        # information, and it in turn may then diseminate sensitive information
-        # through insecure notification channels if so configured.
-        #
-        #sentry:
-        #    dsn: "..."
-
-        # Flags to enable Prometheus metrics which are not suitable to be
-        # enabled by default, either for performance reasons or limited use.
-        #
-        metrics_flags:
-            # Publish synapse_federation_known_servers, a gauge of the number of
-            # servers this homeserver knows about, including itself. May cause
-            # performance problems on large homeservers.
-            #
-            #known_servers: true
-
-        # Whether or not to report anonymized homeserver usage statistics.
-        #
-        """
-
-        if report_stats is None:
-            res += "#report_stats: true|false\n"
-        else:
-            res += "report_stats: %s\n" % ("true" if report_stats else "false")
-
-        res += """
-        # The endpoint to report the anonymized homeserver usage statistics to.
-        # Defaults to https://matrix.org/report-usage-stats/push
-        #
-        #report_stats_endpoint: https://example.com/report-usage-stats/push
-        """
+        res = "report_stats: %s\n" % ("true" if report_stats else "false")
```

Comment:
The comment describing `report_stats` says

https://github.com/matrix-org/synapse/blob/210e8b104b2aed410b080f7a9bdf99cf9891e0cf/synapse/config/_base.py#L446-L447

which I think is no longer accurate with the reported change.

Note that we ask users to explicitly opt-in or out-out of stats reporting when reading config:

https://github.com/matrix-org/synapse/blob/210e8b104b2aed410b080f7a9bdf99cf9891e0cf/synapse/config/_base.py#L827-L830

https://github.com/matrix-org/synapse/blob/210e8b104b2aed410b080f7a9bdf99cf9891e0cf/synapse/config/_base.py#L97-L112

So I think we do want the default config to leave `report_stats` unset---maybe even with the commented-out example as on line -108?

Reply:
I very much want avoid comments, and especially commented out lines. I've changed it so that if `report_stats` is `None`, the config will generate the string "report_stats:" which should force the user to fill it in when they try and run it. Let me know if this is an acceptable change.

Reply:
(sorry, I don't think I explained myself very well.)

Your motivation makes sense, but I think the proposal in f5b16c1 isn't quite right either.

- this now treats `report_stats=False` in the same way as `report_stats=None`
- [`_base.py` looks for the existence of the key `"report_stats"`](https://github.com/matrix-org/synapse/blob/d38d242411b8910dfacde1e61fd3a0ec5cbcaa66/synapse/config/_base.py#L822-L825). But when the yaml syntax `report_stats:` is parsed, that key will be present! Demonstration:
    ```python
    >>> yaml.safe_load("x:")
    {'x': None}
    ```
Let me make a suggestion and see what you make of it...

Reply:
Good point, I have changed it to your suggestion.